include/stdexec/execution.hpp

```diff
@@ -5485,6 +5485,203 @@ namespace stdexec {
   using __when_all::transfer_when_all_with_variant_t;
   inline constexpr transfer_when_all_with_variant_t transfer_when_all_with_variant{};
 
+  namespace __when_any_ {
+  template <class _BaseEnv>
+  using __env_t =
+      __make_env_t<_BaseEnv, __with<get_stop_token_t, in_place_stop_token>>;
+
+  template <class _Receiver, class _Op> struct __receiver_id {
+    class __t {
+    public:
+      using __id = __receiver_id;
+
+      explicit __t(_Op* __op) noexcept : __op_{__op} {}
+
+    private:
+      _Op* __op_;
+
+      template <class _CPO, class... _Args>
+      void notify(_CPO, _Args&&... __args) noexcept {
+        __op_->notify(_CPO{}, ((_Args &&) __args)...);
+      }
+
+      __env_t<env_of_t<_Receiver>> get_env() const noexcept {
+        using with_token = __with<get_stop_token_t, in_place_stop_token>;
+        auto __token = with_token{__op_->__stop_source_.get_token()};
+        return __make_env(stdexec::get_env(__op_->__receiver_), (with_token &&) __token);
+      }
+
+      template <__one_of<set_value_t, set_error_t, set_stopped_t> _CPO,
+                class... _Args>
+      friend void tag_invoke(_CPO, __t&& __self, _Args&&... __args) noexcept {
+        __self.notify(_CPO{}, (_Args &&) __args...);
+      }
+
+      friend __env_t<env_of_t<_Receiver>> tag_invoke(get_env_t, const __t& __self) noexcept {
+        return __self.get_env();
+      }
+    };
+  };
+
+  template <class _R, class _Op> using __receiver = __t<__receiver_id<_R, _Op>>;
+
+  template <class _Env, class...> struct __completion_signatures;
+
+  template <class _Env, class _Sender, class... _Senders>
+  struct __completion_signatures<_Env, _Sender, _Senders...> {
+    using __add_sig_line = __minvoke<__mconcat<__q<completion_signatures>>,
+                                     completion_signatures_of_t<_Senders, _Env>...>;
+
+    using __t = make_completion_signatures<_Sender, _Env, __add_sig_line>;
+  };
+
+  template <class _Env, class... _Senders>
+  using __completion_signatures_t = __t<__completion_signatures<_Env, _Senders...>>;
+
+  template <class _Sig> struct __signature_to_tuple;
+
+  template <class _Ret, class... _Args> struct __signature_to_tuple<_Ret(_Args...)> {
+    using __t = std::tuple<_Ret, _Args...>;
+  };
```

Comment:
I would have to benchmark it to be sure, but I bet something like this would compile quicker:

```c++
template <class _Ret, class... _Args>
  std::tuple<_Ret, _Args...> __signature_to_tuple_(_Ret(*)(_Args...));
template <class _Sig>
  using __signature_to_tuple =
    decltype(stdexec::__signature_to_tuple_((_Sig*) nullptr));
```

Instantiating a class template is expensive.

I bet you could do something similar with the `__completion_signatures` class template above.

Reply:
I have tried to benchmark that with `-ftime-report` both on g++-11 and clang++-16 on a reduced example that uses `exec::when_any`. My benchmark was probably bad but I couldn't find a clear winner. gcc and clang had differing favourites, by small margins. But I will adopt that change and the analouge one for `__completion_signatures` anyway, because the code seems simpler to me.